File: backend/app/schemas/v2/unblock.py
"""詰まり解消相談 API (W-12d BE) Pydantic v2 schemas.

エンドポイント:
  - POST /api/v1/schedule/v2/propose-unblock         (admin/manager, read-only)
  - POST /api/v1/schedule/v2/propose-unblock/apply   (admin/manager, 1 TX 書込)

設計書: docs/plans/unblock-consult-design.md §2.2.

方針:
  - 候補情報は propose-slots と同形 (address/lat/lng + 希望 + 拠点 + 週) を踏襲する。
  - ``time`` は HH:MM 文字列でやり取りする (propose_slots と同一作法)。
  - 全モデル ``extra="forbid"`` (既存 v2 作法)。後方互換のため追加フィールドは default 付き。
"""

from __future__ import annotations

import uuid
from datetime import time as time_cls

from pydantic import BaseModel, ConfigDict, Field, field_validator

from app.schemas.v2.propose_slots import (
    ProposeTimeType,
    ProposeVisitFrequency,
    WeekdayCode,
    _parse_hhmm,
)

# ---------------------------------------------------------------------------
# 候補 (propose-slots と同形の入力)
# ---------------------------------------------------------------------------


class _UnblockCandidateBase(BaseModel):
    """対象患者の希望 (propose-slots と同形) + 対象週 / 拠点.

    ``existing_patient_id`` は apply で対象患者を特定するのに使う (適用には必須)。
    propose-unblock は「個別提案 (プール患者クリック) の候補 0 件」から呼ぶため、
    通常は既存患者 (プール患者) の id が入る。
    """

    model_config = ConfigDict(extra="forbid")

    # --- 候補の座標確定 (address 優先 → lat/lng) ---
    address: str | None = Field(default=None, description="候補患者の住所 (geocode 対象)")
    lat: float | None = Field(default=None, ge=-90.0, le=90.0)
    lng: float | None = Field(default=None, ge=-180.0, le=180.0)

    # --- 候補の希望 (患者マスタ項目に準拠) ---
    service_minutes: int = Field(..., ge=1, le=480, description="訪問サービス時間 (分)")
    time_type: ProposeTimeType | None = Field(default=None)
    preferred_start: str | None = Field(default=None, description="HH:MM")
    preferred_end: str | None = Field(default=None, description="HH:MM")
    preferred_weekdays: list[WeekdayCode] = Field(default_factory=list)
    # visit_frequency / frequency_per_week は探索には未使用だが propose-slots と同形にするため受ける
    # (FE が同じ candidate 情報を送るため。extra=forbid で 422 にしない).
    visit_frequency: ProposeVisitFrequency | None = Field(default=None)
    frequency_per_week: int | None = Field(default=None, ge=1, le=7)
    requires_multiple_staff: bool = Field(default=False)
    sex_restriction: str | None = Field(default=None)
    existing_patient_id: uuid.UUID | None = Field(
        default=None, description="対象患者 (プール患者) の id。propose 時に指定する。"
    )

    # --- 対象週 / 拠点 ---
    iso_year: int = Field(..., ge=2020, le=2100)
    iso_week: int = Field(..., ge=1, le=53)
    office_id: uuid.UUID = Field(..., description="詰まり解消の対象拠点 (office スコープ)")

    @field_validator("preferred_start", "preferred_end")
    @classmethod
    def _validate_hhmm(cls, v: str | None) -> str | None:
        _parse_hhmm(v)  # 形式チェックのみ (実体変換は endpoint 側).
        return v


class ProposeUnblockRequest(_UnblockCandidateBase):
    """``POST /v2/propose-unblock`` リクエスト (read-only)."""

    limit: int = Field(default=5, ge=1, le=20, description="返却プラン数の上限 (上位 N 件)")


# ---------------------------------------------------------------------------
# プラン (レスポンス sub-models)
# ---------------------------------------------------------------------------


class UnblockSlotRef(BaseModel):
    """プラン内の枠参照 (曜日 / コード / 開始時刻).

    apply では simulate レスポンスの plan をそのままエコーバックするため、FE が持たせうる
    追加フィールド (course_label 等) を ``extra="ignore"`` で受け流す (契約の後方互換)。
    """

    model_config = ConfigDict(extra="ignore")

    weekday: int = Field(..., ge=0, le=6, description="0=Mon..6=Sun")
    course_code: str = Field(..., description="A..E / M..M9")
    course_label: str | None = Field(default=None, description="拠点+コード (BE 未送出なら null)")
    start_time: str = Field(..., description="HH:MM")


class UnblockMoveItem(BaseModel):
    """既存訪問を 1 手ずらす (退避)."""

    # ``from`` は予約語のため ``from_`` を alias で受ける (populate_by_name で内部名も許容).
    # plan は apply でエコーバックされるため extra=ignore で FE 追加フィールドを受け流す.
    model_config = ConfigDict(extra="ignore", populate_by_name=True)

    patient_id: uuid.UUID
    patient_name: str
    from_: UnblockSlotRef = Field(..., alias="from")
    to: UnblockSlotRef
    delta_minutes: int | None = Field(
        default=None, description="退避で増える移動 (分/週). 正 = 乱れ増 / 負 = 改善."
    )
    within_preference: bool = Field(
        default=False, description="退避先が患者の希望訪問スケジュール範囲内か"
    )


class UnblockInsertItem(BaseModel):
    """対象患者を配置する枠 (2 名体制なら partner_course_code に相方コース)."""

    model_config = ConfigDict(extra="ignore")

    weekday: int = Field(..., ge=0, le=6, description="0=Mon..6=Sun")
    course_code: str = Field(..., description="A..E / M..M9")
    course_label: str | None = Field(default=None, description="拠点+コード (BE 未送出なら null)")
    start_time: str = Field(..., description="HH:MM")
    end_time: str = Field(..., description="HH:MM")
    partner_course_code: str | None = Field(
        default=None, description="2 名体制ペアの相方 (slot1) コース。通常配置は null。"
    )
    partner_course_label: str | None = Field(default=None, description="相方コースの表示ラベル")


class UnblockPlanItem(BaseModel):
    """開通プラン 1 件 (moves を適用 → insert に対象を配置).

    ``plan_id`` は ``{対象患者id}:{内容ハッシュ}`` 形式の決定的文字列。FE はこれを不透明な
    キーとして round-trip し、apply は先頭の対象患者 id を取り出して配置対象を復元する
    (apply 契約が candidate 情報を持たないため。§2.2 の plan エコーバックだけで自足させる)。
    """

    model_config = ConfigDict(extra="ignore")

    plan_id: str = Field(..., description="{対象患者id}:{内容ハッシュ} の決定的文字列")
    moves: list[UnblockMoveItem] = Field(default_factory=list)
    insert: UnblockInsertItem
    total_delta_minutes: int = Field(default=0, description="全手の delta 合計 (分/週)")
    moved_count: int = Field(default=0, ge=0, description="動かす既存訪問の数 (= len(moves))")


class UnblockUnmovableSummary(BaseModel):
    """動かせない事情の内訳 (N-6「黙って諦めない」). real-blocker 単位のカウント."""

    model_config = ConfigDict(extra="forbid")

    pinned: int = Field(default=0, ge=0, description="ピン留めで動かせない real-blocker 数")
    locked: int = Field(default=0, ge=0, description="可動域=完全固定で動かせない数")
    two_staff: int = Field(default=0, ge=0, description="2 名体制で v1 は動かさない数")
    pair: int = Field(default=0, ge=0, description="同住所ペアの片割れで動かさない数")
    dismissed: int = Field(default=0, ge=0, description="却下記憶で退避を出さなかった数")
    confirmation_required: int = Field(
        default=0,
        ge=0,
        description="退避先が希望外+movability不明で患者確認が要るため出さなかった数",
    )


class ProposeUnblockResponse(BaseModel):
    """``POST /v2/propose-unblock`` レスポンス (read-only)."""

    model_config = ConfigDict(extra="forbid")

    plans: list[UnblockPlanItem] = Field(default_factory=list)
    unmovable_summary: UnblockUnmovableSummary = Field(default_factory=UnblockUnmovableSummary)
    state_token: str = Field(..., description="apply 用の楽観ロック指紋 (sha256, office スコープ)")


# ---------------------------------------------------------------------------
# apply
# ---------------------------------------------------------------------------


class ProposeUnblockApplyRequest(BaseModel):
    """``POST /v2/propose-unblock/apply`` リクエスト (1 TX / all-or-nothing).

    ``plan`` は simulate レスポンスの 1 件をそのままエコーバックする。
    ``target_patient_id`` は明示フィールドとして必須 (plan_id 先頭の UUID と照合し、
    さらにサーバが plan 内容から plan_id を再導出してハッシュ指紋の一致を検証する)。
    ``state_token`` はサーバが再計算し不一致なら 409 (詰まり状況が変わった = 再探索が必要)。
    """

    model_config = ConfigDict(extra="forbid")

    office_id: uuid.UUID
    iso_year: int = Field(..., ge=2020, le=2100)
    iso_week: int = Field(..., ge=1, le=53)
    plan: UnblockPlanItem
    state_token: str = Field(..., min_length=1)
    target_patient_id: uuid.UUID = Field(
        ...,
        description="配置対象患者の UUID (plan_id 先頭との一致 + ハッシュ再導出で改竄を検出).",
    )


class ProposeUnblockApplyResponse(BaseModel):
    """``POST /v2/propose-unblock/apply`` レスポンス."""

    model_config = ConfigDict(extra="forbid")

    applied_moves: int = Field(..., ge=0, description="適用した退避手数 (= len(plan.moves))")
    inserted: bool = Field(..., description="対象患者を配置できたか")
    warnings: list[str] = Field(
        default_factory=list,
        description="N-4 再検証で検出した非致命の警告 (現場向け日本語). ブロックしない.",
    )


def parse_hhmm_or_none(value: str | None) -> time_cls | None:
    """HH:MM を time に変換する薄いラッパ (endpoint から再利用)."""
    return _parse_hhmm(value)


__all__ = [
    "ProposeUnblockApplyRequest",
    "ProposeUnblockApplyResponse",
    "ProposeUnblockRequest",
    "ProposeUnblockResponse",
    "UnblockInsertItem",
    "UnblockMoveItem",
    "UnblockPlanItem",
    "UnblockSlotRef",
    "UnblockUnmovableSummary",
]

File: frontend/lib/schemas/v2/unblock.ts
/**
 * W-12d 詰まり解消相談 (propose-unblock) zod スキーマ。
 *
 * BE 契約 (設計書 `docs/plans/unblock-consult-design.md` §2.2 を正典とする):
 *   POST /api/v1/schedule/v2/propose-unblock       (read-only 探索)
 *   POST /api/v1/schedule/v2/propose-unblock/apply  (1 TX 適用)
 *   RBAC: admin / manager (BE 側で担保)。
 *
 * BE は並行実装中のため、契約を正として寛容パース (.default/.catch/nullish) で
 * 先行実装する (propose_slots / poolBulk の後方互換規約を踏襲)。
 *
 * リクエストの candidate 情報は propose-slots と同形 (address/lat/lng/希望条件 +
 * existing_patient_id) + { office_id, iso_year, iso_week, limit?: 5 }。
 */
import { z } from 'zod';

import { weekdayCodeSchema } from './board';
import { proposeTimeTypeSchema, proposeVisitFrequencySchema } from './propose_slots';

// ---------------------------------------------------------------------------
// 寛容パースヘルパー (未知要素をセクション全滅させず静かに除外)
// ---------------------------------------------------------------------------

/** 配列を要素単位で safeParse し、パース不能な要素を除外する寛容配列スキーマ。 */
function tolerantArray<T extends z.ZodTypeAny>(schema: T) {
  return z.preprocess((val) => {
    if (!Array.isArray(val)) return val;
    return val.filter((el) => schema.safeParse(el).success);
  }, z.array(schema));
}

// ---------------------------------------------------------------------------
// Request
// ---------------------------------------------------------------------------

/**
 * POST /v2/propose-unblock リクエスト。
 * candidate 情報は propose-slots と同形。office_id は単一 (state_token が拠点単位の
 * 指紋のため)。limit は既定 5 (上位プランのみ)。
 */
export const proposeUnblockRequestSchema = z.object({
  // 候補の座標確定 (address 優先 → lat/lng)。
  address: z.string().nullish(),
  lat: z.number().min(-90).max(90).nullish(),
  lng: z.number().min(-180).max(180).nullish(),

  // 候補の希望 (患者マスタ項目に準拠)。
  service_minutes: z.number().int().min(1).max(480),
  time_type: proposeTimeTypeSchema.nullish(),
  preferred_start: z.string().nullish(),
  preferred_end: z.string().nullish(),
  preferred_weekdays: z.array(weekdayCodeSchema).default([]),
  visit_frequency: proposeVisitFrequencySchema.nullish(),
  frequency_per_week: z.number().int().min(1).max(7).nullish(),
  requires_multiple_staff: z.boolean().default(false),
  sex_restriction: z.string().nullish(),
  existing_patient_id: z.string().uuid().nullish(),

  // 対象週 / 拠点 (単一)。
  // W-13a: office_id は省略可能。省略時 BE は対象患者の主担当拠点 (primary_office_id)
  // から自動解決する。ページで拠点選択中はそれを送る (従来どおり)。
  office_id: z.string().uuid().nullish(),
  iso_year: z.number().int().min(2020).max(2100),
  iso_week: z.number().int().min(1).max(53),

  limit: z.number().int().min(1).max(20).default(5),
});
export type ProposeUnblockRequest = z.input<typeof proposeUnblockRequestSchema>;

// ---------------------------------------------------------------------------
// Response sub-models
// ---------------------------------------------------------------------------

/** 手の始点 / 終点 (曜日・コース・時刻)。course_label は BE 未送出なら code で代替。 */
export const unblockMoveEndpointSchema = z.object({
  weekday: z.number().int().min(0).max(6),
  course_code: z.string(),
  course_label: z.string().nullish(),
  start_time: z.string(),
});
export type UnblockMoveEndpoint = z.infer<typeof unblockMoveEndpointSchema>;

/** 既存訪問 1 件の退避 (from → to)。within_preference=true は「希望範囲内」の手。 */
export const unblockMoveSchema = z.object({
  patient_id: z.string(),
  patient_name: z.string(),
  from: unblockMoveEndpointSchema,
  to: unblockMoveEndpointSchema,
  /** 手の効果 (分/週)。compute_exact_marginal。未評価は null。 */
  delta_minutes: z.number().nullish(),
  within_preference: z.boolean().default(false),
});
export type UnblockMove = z.infer<typeof unblockMoveSchema>;

/** 対象患者の配置先 (この枠に入れる)。partner_* は 2名体制ペアのときのみ。 */
export const unblockInsertSchema = z.object({
  weekday: z.number().int().min(0).max(6),
  course_code: z.string(),
  course_label: z.string().nullish(),
  start_time: z.string(),
  end_time: z.string().nullish(),
  /** 2名体制ペアの相方コース (非 null ならペア配置)。 */
  partner_course_code: z.string().nullish(),
  partner_course_label: z.string().nullish(),
});
export type UnblockInsert = z.infer<typeof unblockInsertSchema>;

/**
 * W-13b: 影響コースの before/after スナップショット 1 件 (BE が plan ごとに返す)。
 * entries は improvement 系スナップショット (CourseSnapshotVisit) と同形。
 * before/after を CourseMoveTimeline 系の正典部品で「変更前 → 変更後」の 2 列表示する。
 * 旧 BE (courses 未送出) でも壊れないよう寛容パース。
 */
export const unblockCourseEntrySchema = z.object({
  patient_id: z.string(),
  patient_name: z.string(),
  start_time: z.string(),
  end_time: z.string(),
});
export type UnblockCourseEntry = z.infer<typeof unblockCourseEntrySchema>;

export const unblockCourseBeforeAfterSchema = z.object({
  office_name: z.string().nullish(),
  weekday: z.number().int().min(0).max(6),
  course_code: z.string(),
  course_label: z.string().nullish(),
  before: tolerantArray(unblockCourseEntrySchema).default([]),
  after: tolerantArray(unblockCourseEntrySchema).default([]),
  // イベント表示 (2026-07-27): 担当スタッフのイベント (before/after 共通・緑カード)。
  events: z
    .array(z.object({ title: z.string(), start_time: z.string(), end_time: z.string() }))
    .catch([]),
});
export type UnblockCourseBeforeAfter = z.infer<typeof unblockCourseBeforeAfterSchema>;

/** 開通プラン 1 件 (moves ≤ 2 + insert 1)。乱れの小さい順にランキング済み。 */
export const unblockPlanSchema = z.object({
  plan_id: z.string(),
  moves: z.array(unblockMoveSchema).default([]),
  insert: unblockInsertSchema,
  total_delta_minutes: z.number().default(0),
  moved_count: z.number().int().default(0),
  // W-13b: 影響する全コース (移動元・移動先・配置先) の before/after。旧 BE 未送出 → []。
  courses: tolerantArray(unblockCourseBeforeAfterSchema).default([]),
  // W-15: 定員起因 (capacity_full) の探索で、ブロッカーを他コースへ退避させて定員を
  // 空けるプランなら true。FE は小バッジ「定員内に収まります」で明示する。
  // 旧 BE / 時間起因プラン (未送出) → false (寛容パース)。
  frees_capacity: z.boolean().default(false),
});
export type UnblockPlan = z.infer<typeof unblockPlanSchema>;

/** 動かせない訪問の内訳会計 (N-6「黙って諦めない」)。全欠落は 0 に落とす。 */
export const unblockUnmovableSummarySchema = z.object({
  pinned: z.number().int().catch(0).default(0),
  locked: z.number().int().catch(0).default(0),
  two_staff: z.number().int().catch(0).default(0),
  pair: z.number().int().catch(0).default(0),
  dismissed: z.number().int().catch(0).default(0),
  confirmation_required: z.number().int().catch(0).default(0),
});
export type UnblockUnmovableSummary = z.infer<typeof unblockUnmovableSummarySchema>;

const EMPTY_UNMOVABLE: UnblockUnmovableSummary = {
  pinned: 0,
  locked: 0,
  two_staff: 0,
  pair: 0,
  dismissed: 0,
  confirmation_required: 0,
};

/** POST /v2/propose-unblock レスポンス。 */
export const proposeUnblockResponseSchema = z.object({
  plans: tolerantArray(unblockPlanSchema).default([]),
  unmovable_summary: unblockUnmovableSummarySchema.catch(EMPTY_UNMOVABLE).default(EMPTY_UNMOVABLE),
  /** 楽観ロック用指紋 (apply で再計算し不一致なら 409)。scope-opt と同じ PFV 指紋。 */
  state_token: z.string(),
  /**
   * W-13a: office_id 省略時に BE が主担当拠点から解決した拠点 ID (apply の office スコープに使う)。
   * 旧 BE 未送出 → null。FE は apply 時 officeId ?? resolved_office_id の順で解決する。
   */
  resolved_office_id: z.string().uuid().nullish(),
});
export type ProposeUnblockResponse = z.infer<typeof proposeUnblockResponseSchema>;

// ---------------------------------------------------------------------------
// apply: Request / Response
// ---------------------------------------------------------------------------

/**
 * POST /v2/propose-unblock/apply リクエスト (設計書 §2.2)。
 *   - plan: propose-unblock の plan をそのまま送る。
 *   - target_patient_id: 対象患者 UUID。BE が plan_id 内の患者 ID との一致＋指紋照合で 422 を返す。
 *   - state_token 不一致は BE が 409 (FE は再探索導線)。
 *   - 反映先は pattern_and_week 固定 (v1・P-6)。change_scope は持たせない。
 */
export const unblockApplyRequestSchema = z.object({
  // W-13a: office_id 省略時は BE が対象患者の主担当拠点から再解決する (探索と同じ挙動)。
  office_id: z.string().nullish(),
  iso_year: z.number().int(),
  iso_week: z.number().int().min(1).max(53),
  /** 対象患者 UUID (BE の指紋照合に使用)。 */
  target_patient_id: z.string().uuid(),
  plan: unblockPlanSchema,
  state_token: z.string(),
});
export type UnblockApplyRequest = z.input<typeof unblockApplyRequestSchema>;

/** POST /v2/propose-unblock/apply レスポンス (設計書 §2.2)。 */
export const unblockApplyResponseSchema = z.object({
  applied_moves: z.number().int().default(0),
  inserted: z.boolean().default(false),
  warnings: z.array(z.string()).catch([]),
});
export type UnblockApplyResponse = z.infer<typeof unblockApplyResponseSchema>;
